### DISCUSSION TITLE
Build SoftEther for windows x64

### DISCUSSION THREAD

**Discussion by ngstargate101** [SoftEtherVPN/SoftEtherVPN#1438] [Q&A] [answered]
Hello.
I'm try build SoftEther VPN for Windows x64 using  "/src/BUILD_WINDOWS.md". But I cannot build - got this error
Error  C1189  #error:  "This code requires at least SSE2."
..\SoftEtherVPN-1\3rdparty\BLAKE2\sse\blake2-config.h

Can you help me?

**Comment by andrewfer000** [comment]
What is your CPU? It needs to support SSE2 instructions.

**Reply by ngstargate101** [reply]
Intel Xeon E3-1231v3
Instructions sets: MMX, SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2, EM64T, AES, AVX, AVX2, FMA3

**Reply by davidebeatrici** [reply]
Unfortunately MSVC doesn't set `__SSE2__`.

The solution consists in adding

```cmake
if(HAS_SSE2)
  target_compile_definitions(cedar PRIVATE "__SSE2__")
endif()
```

below

https://github.com/SoftEtherVPN/SoftEtherVPN/blob/bf14817f1f09f8b8107b13add08f483ad99abb7c/src/Cedar/CMakeLists.txt#L24

**Reply by ngstargate101** [reply]
I place this code in SoftEtherVPN/src/Cedar/CMakeLists.txt and got 84 errors...
This is first 9.
Error	C2065	'AddrModeFlat': undeclared identifier	\SoftEther\src\Mayaqua\Win32.c	2617	
Error	C2065	'AddrModeFlat': undeclared identifier	\SoftEther\src\Mayaqua\Win32.c	2618	
Error	C2065	'AddrModeFlat': undeclared identifier	\SoftEther\src\Mayaqua\Win32.c	2619	
Error	C2065	'IMAGEHLP_LINE64': undeclared identifier	\SoftEther\src\Mayaqua\Win32.c	2686	
Error	C2065	'IMAGEHLP_SYMBOL64': undeclared identifier	\SoftEther\src\Mayaqua\Win32.c	2685	
Error	C2065	'IMAGEHLP_SYMBOL64': undeclared identifier	\SoftEther\src\Mayaqua\Win32.c	2701	
Error	C2065	'line': undeclared identifier	\SoftEther\src\Mayaqua\Win32.c	2686	
Error	C2065	'line': undeclared identifier	\SoftEther\src\Mayaqua\Win32.c	2715	
Error	C2065	'line': undeclared identifier	\SoftEther\src\Mayaqua\Win32.c	2716

**Reply by davidebeatrici** [reply]
That's unexpected.

What's the version of the Windows SDK and compiler you're using?

**Reply by ngstargate101** [reply]
From BUILD_WINDOWS.md
Requirements:

1. Download Visual Studio 2019 (Community Edition is fine).
2. During install, make sure to check "Desktop development with C++" under "Workloads".
3. Click on individual components and scroll until you see "Visual C++ tools for CMake" under the compilers section. Make sure this is checked.
4. Proceed with and finish Visual Studio 2019 installation.
5. Install the needed submodules to build the project, avoiding CMake telling you to do so with: `git submodule update --init --recursive`

**Reply by davidebeatrici** [reply]
What are the exact commands you're using to build the project?

**Reply by ngstargate101** [reply]
From BUILD_WINDOWS.md
Building:

Launch Visual Studio.
Go to the File menu click Open --> CMake. Then navigate to where you cloned the project and open the CMakeLists.txt file in the projects root directory.

Visual Studio will proceed to start the CMake configuration process and once its finished, without errors.
Go to toolbar and click Build -> Build All

Previous version 5.01.9674 without module BLAKE2 builded ok.

**Comment by domosekai** [comment]
Very good question. The current BUILD_WINDOWS.md seems outdated... 
When I tried to build with VS 2019 Community last time, I gave up after 10+ hours of trial and error. I ended up creating a draft PR and making use of the Azure pipeline...

Edited: I finally made it. Here's how I did in addition to the procedures in BUILD_WINDOWS.md.

1. Install vcpkg
2. Integrate vcpkg with vs 2019: `vcpkg integrate install`
2. Install dependencies in vcpkg: `vcpkg install libsodium openssl zlib --triplet=x64-windows-static-md`
5. Change configuration type to RelWithDebInfo or Release (otherwise you will get those undeclared identifier errors)
4. Add the lines according to @davidebeatrici 's reply if you use the default MSVC compiler
5. Set `BUILD_NUMBER` in cmake settings

Alternatively, you can use clang compiler instead of MSVC
1. In vs installer, install C++ clang compiler
2. Change cmake toolset to `clang_cl_x64_x64`
4. In cmake variables, click *Show advanced variables* and change `CMAKE_CXX_COMPILER` to clang-cl.exe

**Comment by domosekai** [comment, marked as answer]
The build instruction for Windows has been rewritten. Please follow the new guide.

**Reply by isurunuwanthilaka** [reply]
Hi , how can I set up SoftEther on MacOS, is there a guide?
